util/logutil/inconsist/reporter.go

```diff
@@ -0,0 +1,264 @@
+	keyLocation, err := tikvStore.GetRegionCache().LocateKey(tikv.NewBackofferWithVars(context.Background(), 500, nil), encodedKey)
+	if err != nil {
+		return 0
+	}
+	return keyLocation.Region.GetID()
+}
+
+// Reporter is a helper to generate report.
+type Reporter struct {
+	HandleEncode func(handle kv.Handle) kv.Key
+	IndexEncode  func(idxRow *RecordData) kv.Key
+	Tbl          *model.TableInfo
+	Idx          *model.IndexInfo
+	Sctx         sessionctx.Context
+}
+
+func (r *Reporter) decodeRowMvccData(_ kv.Key, respValue *kvrpcpb.MvccGetByKeyResponse, outMap map[string]interface{}) {
+	colMap := make(map[int64]*types.FieldType, 3)
+	for _, col := range r.Tbl.Columns {
+		colMap[col.ID] = &col.FieldType
+	}
+
+	if respValue.Info != nil {
+		var err error
+		datas := make(map[string][]map[string]string)
+		for _, w := range respValue.Info.Writes {
+			if len(w.ShortValue) > 0 {
+				datas[strconv.FormatUint(w.StartTs, 10)], err = decodeMvccData(w.ShortValue, colMap, r.Tbl)
+			}
+		}
+
+		for _, v := range respValue.Info.Values {
+			if len(v.Value) > 0 {
+				datas[strconv.FormatUint(v.StartTs, 10)], err = decodeMvccData(v.Value, colMap, r.Tbl)
+			}
+		}
+		if len(datas) > 0 {
+			outMap["decoded"] = datas
+			if err != nil {
+				outMap["decode_error"] = err.Error()
+			}
+		}
+	}
+}
+
+func (r *Reporter) decodeIndexMvccData(key kv.Key, respValue *kvrpcpb.MvccGetByKeyResponse, outMap map[string]interface{}) {
+	if respValue.Info != nil {
+		var (
+			hd    kv.Handle
+			err   error
+			datas = make(map[string]map[string]string)
+		)
+		for _, w := range respValue.Info.Writes {
+			if len(w.ShortValue) > 0 {
+				hd, err = tablecodec.DecodeIndexHandle(key, w.ShortValue, len(r.Idx.Columns))
+				if err == nil {
+					datas[strconv.FormatUint(w.StartTs, 10)] = map[string]string{"handle": hd.String()}
+				}
+			}
+		}
+		for _, v := range respValue.Info.Values {
+			if len(v.Value) > 0 {
+				hd, err = tablecodec.DecodeIndexHandle(key, v.Value, len(r.Idx.Columns))
+				if err == nil {
+					datas[strconv.FormatUint(v.StartTs, 10)] = map[string]string{"handle": hd.String()}
+				}
+			}
+		}
+		if len(datas) > 0 {
+			outMap["decoded"] = datas
+			if err != nil {
+				outMap["decode_error"] = err.Error()
+			}
+		}
+	}
+}
+
+func decodeMvccData(bs []byte, colMap map[int64]*types.FieldType, tb *model.TableInfo) ([]map[string]string, error) {
+	rs, err := tablecodec.DecodeRowToDatumMap(bs, colMap, time.UTC)
+	var record []map[string]string
+	for _, col := range tb.Columns {
+		if c, ok := rs[col.ID]; ok {
+			data := "nil"
+			if !c.IsNull() {
+				data, err = c.ToString()
+			}
+			record = append(record, map[string]string{col.Name.O: data})
+		}
+	}
+	return record, err
+}
+
+// ReportLookupInconsistent reports inconsistent when index rows is more than record rows.
+func (r *Reporter) ReportLookupInconsistent(ctx context.Context, idxCnt, tblCnt int, missHd, fullHd []kv.Handle, missRowIdx []RecordData) error {
```

Comment:
Do we need to save the reported inconsistency errors into a inner history physical table as the logs could be rotated or cleaned, so that we could know the error history, and generally this list will not be too long.

Reply:
good idea, but it seems we have no persistent log table now( except "test version audit log"), maybe a little complex and suitable for doing it in other PR to make a persistent log to support more general function..

maybe it's event table ---- show critical event like reboot or critical questions like this and very suitable to display it in dashboard in production view.

but write file log and report error is first step, maybe final one will:

- report error
- write file log
- write some metrics to alert
- write event table like above